File: Modules/module_jetpack/src/main/java/com/example/william/my/module/jetpack/model/LiveDataViewModel.java
package com.example.william.my.module.jetpack.model;

import androidx.arch.core.util.Function;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;

public class LiveDataViewModel extends ViewModel {

    private final MutableLiveData<Integer> _likes;

    private final LiveData<String> likes;

    public LiveDataViewModel() {
        this._likes = new MutableLiveData<>(0);
        this.likes = Transformations.map(_likes, new Function<Integer, String>() {
            @Override
            public String apply(Integer input) {
                return String.valueOf(input);
            }
        });
    }

    public LiveData<String> getLikes() {
        return likes;
    }

    public void onLike() {
        _likes.setValue(_likes.getValue() + 1);
    }
}

File: modules/module_sample/src/main/kotlin/com/example/william/my/module/sample/model/LiveDataViewModel.kt
package com.example.william.my.module.sample.model

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.example.william.my.bean.repo.DataRepository

class LiveDataViewModel(private val dataSource: DataRepository) : ViewModel() {

    private val mMutableLiveData = MutableLiveData<Int>()

    val article = Transformations.switchMap(mMutableLiveData) { input ->
        dataSource.getArticle(input)
    }

    fun queryArticle(page: Int) {
        mMutableLiveData.postValue(page)
    }
}

/**
 * 自定义实例，多参构造
 * Factory for [DataBindingViewModel].
 */
object ArticleVMFactory : ViewModelProvider.Factory {

    private val dataSource = DataRepository.getInstance()

    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        @Suppress("UNCHECKED_CAST")
        return LiveDataViewModel(dataSource) as T
    }
}
